File: web/src/validators.ts
import CheckStatus from "./types/CheckStatus";
import { createRegex } from "./utils";

// TODO: array minlen, maxlen, minval, maxval

type ValidatorFunction = (
	field: string,
	value: any,
	raw: string,
	required: any,
	status: CheckStatus,
) => boolean;

const validators: Record<string, Record<string, ValidatorFunction>> = {
	bool: {},
	int: {
		min,
		max,
	},
	num: {
		min,
		max,
	},
	date: {
		min: minDate,
		max: maxDate,
	},
	string: {
		minlen,
		maxlen,
		pattern,
	},
	array: {
		minlen: minlenArray,
		maxlen: maxlenArray,
		unique: unique,
	},
};
export default validators;

function min(field: string, value: number, raw: string, required: number, status: CheckStatus) {
	if (value < required) {
		status.errors.push({
			path: [...status.path],
			message: `'${field}' must be at least ${raw}`,
		});
		return false;
	}
	return true;
}

function max(field: string, value: number, raw: string, required: number, status: CheckStatus) {
	if (value > required) {
		status.errors.push({
			path: [...status.path],
			message: `'${field}' cannot be more than ${raw}`,
		});
		return false;
	}
	return true;
}

function minDate(field: string, value: Date, raw: string, required: Date, status: CheckStatus) {
	if (value < required) {
		status.errors.push({
			path: [...status.path],
			message: `'${field}' must be at least ${raw}`,
		});
		return false;
	}
	return true;
}

function maxDate(field: string, value: Date, raw: string, required: Date, status: CheckStatus) {
	if (value > required) {
		status.errors.push({
			path: [...status.path],
			message: `'${field}' cannot be after ${raw}`,
		});
		return false;
	}
	return true;
}

function minlen(field: string, value: string, raw: string, required: number, status: CheckStatus) {
	if (value.length < required) {
		status.errors.push({
			path: [...status.path],
			message: `'${field}' must be at least ${raw} characters`,
		});
		return false;
	}
	return true;
}

function maxlen(field: string, value: string, raw: string, required: number, status: CheckStatus) {
	if (value.length > required) {
		status.errors.push({
			path: [...status.path],
			message: `'${field}' cannot be more than ${raw} characters`,
		});
		return false;
	}
	return true;
}

function pattern(field: string, value: string, raw: string, required: string, status: CheckStatus) {
	const regex = createRegex(required);
	if (regex === undefined) {
		// This should never happen...
		status.errors.push({
			path: [...status.path],
			message: `Unsupported pattern for '${field}': ${raw}`,
		});
		return false;
	}
	if (!regex.test(value)) {
		status.errors.push({
			path: [...status.path],
			message: `'${field}' doesn't match pattern '${raw}'`,
		});
		return false;
	}
	return true;
}
function minlenArray(
	field: string,
	value: Array<any>,
	raw: string,
	required: number,
	status: CheckStatus,
) {
	if (value.length < required) {
		status.errors.push({
			path: [...status.path],
			message: `'${field}' must contain at least ${raw} items`,
		});
		return false;
	}
	return true;
}

function maxlenArray(
	field: string,
	value: Array<any>,
	raw: string,
	required: number,
	status: CheckStatus,
) {
	if (value.length > required) {
		status.errors.push({
			path: [...status.path],
			message: `'${field}' cannot contain more than ${raw} items`,
		});
		return false;
	}
	return true;
}

function unique(
	field: string,
	value: Array<any>,
	_raw: string,
	_required: number,
	status: CheckStatus,
) {
	// Maybe only use a Set if longer than a certain length?
	let set = new Set();
	let ok = true;
	for (let i = 0; i < value.length; i++) {
		if (set.has(value[i])) {
			status.errors.push({
				path: [...status.path],
				message: `'${field}' value '${value[i]}' is not unique`,
			});
			ok = false;
		} else {
			set.add(value[i]);
		}
	}
	return ok;
}

File: dotnet/Sdnx.Core/Validators.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;
using Sdnx.Core;

namespace Sdnx.Core
{
    public delegate bool ValidatorFunction(
        string field,
        object? value,
        string raw,
        object? required,
        CheckStatus status
    );

    public static class Validators
    {
        // TODO: array minval, maxval

        public static readonly Dictionary<string, Dictionary<string, ValidatorFunction>> All = new()
        {
            ["bool"] = new Dictionary<string, ValidatorFunction>(),
            ["int"] = new Dictionary<string, ValidatorFunction>
            {
                ["min"] = Min,
                ["max"] = Max
            },
            ["num"] = new Dictionary<string, ValidatorFunction>
            {
                ["min"] = Min,
                ["max"] = Max
            },
            ["date"] = new Dictionary<string, ValidatorFunction>
            {
                ["min"] = MinDate,
                ["max"] = MaxDate
            },
            ["string"] = new Dictionary<string, ValidatorFunction>
            {
                ["minlen"] = Minlen,
                ["maxlen"] = Maxlen,
                ["pattern"] = Pattern
            },
            ["array"] = new Dictionary<string, ValidatorFunction>
            {
                ["minlen"] = MinlenArray,
                ["maxlen"] = MaxlenArray,
                ["unique"] = Unique
            }
        };

        public static bool Min(string field, object? value, string raw, object? required, CheckStatus status)
        {
            double numValue = 0;
            if (value is int intVal)
            {
                numValue = intVal;
            }
            else if (value is double doubleVal)
            {
                numValue = doubleVal;
            }
            else
            {
                return true;
            }

            double requiredValue = 0;
            if (required is int reqInt)
            {
                requiredValue = reqInt;
            }
            else if (required is double reqDouble)
            {
                requiredValue = reqDouble;
            }
            else
            {
                return true;
            }

            if (numValue < requiredValue)
            {
                status.Errors.Add(new CheckError(new List<string>(status.Path), $"'{field}' must be at least {raw}"));
                return false;
            }
            return true;
        }

        public static bool Max(string field, object? value, string raw, object? required, CheckStatus status)
        {
            double numValue = 0;
            if (value is int intVal)
            {
                numValue = intVal;
            }
            else if (value is double doubleVal)
            {
                numValue = doubleVal;
            }
            else
            {
                return true;
            }

            double requiredValue = 0;
            if (required is int reqInt)
            {
                requiredValue = reqInt;
            }
            else if (required is double reqDouble)
            {
                requiredValue = reqDouble;
            }
            else
            {
                return true;
            }

            if (numValue > requiredValue)
            {
                status.Errors.Add(new CheckError(new List<string>(status.Path), $"'{field}' cannot be more than {raw}"));
                return false;
            }
            return true;
        }

        public static bool MinDate(string field, object? value, string raw, object? required, CheckStatus status)
        {
            if (value is not DateTime dateValue)
            {
                return true;
            }

            DateTime requiredDate;
            if (required is DateTime reqDate)
            {
                requiredDate = reqDate;
            }
            else if (required is string reqStr && DateTime.TryParse(reqStr, out var parsedDate))
            {
                requiredDate = parsedDate;
            }
            else
            {
                return true;
            }

            if (dateValue < requiredDate)
            {
                status.Errors.Add(new CheckError(new List<string>(status.Path), $"'{field}' must be at least {raw}"));
                return false;
            }
            return true;
        }

        public static bool MaxDate(string field, object? value, string raw, object? required, CheckStatus status)
        {
            if (value is not DateTime dateValue)
            {
                return true;
            }

            DateTime requiredDate;
            if (required is DateTime reqDate)
            {
                requiredDate = reqDate;
            }
            else if (required is string reqStr && DateTime.TryParse(reqStr, out var parsedDate))
            {
                requiredDate = parsedDate;
            }
            else
            {
                return true;
            }

            if (dateValue > requiredDate)
            {
                status.Errors.Add(new CheckError(new List<string>(status.Path), $"'{field}' cannot be after {raw}"));
                return false;
            }
            return true;
        }

        public static bool Minlen(string field, object? value, string raw, object? required, CheckStatus status)
        {
            if (value is not string strValue)
            {
                return true;
            }

            int requiredValue = 0;
            if (required is int reqInt)
            {
                requiredValue = reqInt;
            }
            else if (required is double reqDouble)
            {
                requiredValue = (int)reqDouble;
            }
            else
            {
                return true;
            }

            if (strValue.Length < requiredValue)
            {
                status.Errors.Add(new CheckError(new List<string>(status.Path), $"'{field}' must be at least {raw} characters"));
                return false;
            }
            return true;
        }

        public static bool Maxlen(string field, object? value, string raw, object? required, CheckStatus status)
        {
            if (value is not string strValue)
            {
                return true;
            }

            int requiredValue = 0;
            if (required is int reqInt)
            {
                requiredValue = reqInt;
            }
            else if (required is double reqDouble)
            {
                requiredValue = (int)reqDouble;
            }
            else
            {
                return true;
            }

            if (strValue.Length > requiredValue)
            {
                status.Errors.Add(new CheckError(new List<string>(status.Path), $"'{field}' cannot be more than {raw} characters"));
                return false;
            }
            return true;
        }

        public static bool Pattern(string field, object? value, string raw, object? required, CheckStatus status)
        {
            if (value is string strValue && required is string patternStr)
            {
                var regex = Utils.CreateRegex(patternStr);
                if (regex == null)
                {
                    // This should never happen...
                    status.Errors.Add(new CheckError(new List<string>(status.Path), $"Unsupported pattern for '{field}': {raw}"));
                    return false;
                }

                if (!regex.IsMatch(strValue))
                {
                    status.Errors.Add(new CheckError(new List<string>(status.Path), $"'{field}' doesn't match pattern '{raw}'"));
                    return false;
                }
            }
            return true;
        }

        public static bool MinlenArray(string field, object? value, string raw, object? required, CheckStatus status)
        {
            if (value is Array arrValue && required is double requiredValue)
            {
                if (arrValue.Length < requiredValue)
                {
                    status.Errors.Add(new CheckError(new List<string>(status.Path), $"'{field}' must contain at least {raw} items"));
                    return false;
                }
            }
            return true;
        }

        public static bool MaxlenArray(string field, object? value, string raw, object? required, CheckStatus status)
        {
            if (value is Array arrValue && required is double requiredValue)
            {
                if (arrValue.Length > requiredValue)
                {
                    status.Errors.Add(new CheckError(new List<string>(status.Path), $"'{field}' cannot contain more than {raw} items"));
                    return false;
                }
            }
            return true;
        }

        public static bool Unique(string field, object? value, string raw, object? required, CheckStatus status)
        {
            // Maybe only use a Set if longer than a certain length?
            if (value is Array arrValue)
            {
                var set = new HashSet<object?>();
                bool ok = true;

                foreach (var item in arrValue)
                {
                    if (set.Contains(item))
                    {
                        status.Errors.Add(new CheckError(new List<string>(status.Path), $"'{field}' value '{item}' is not unique"));
                        ok = false;
                    }
                    else
                    {
                        set.Add(item);
                    }
                }

                return ok;
            }
            return true;
        }
    }
}
